Marker/Marker/Parser/MarkdownParser.swift

```diff
@@ -22,54 +22,97 @@ internal struct MarkdownParser {
+    private static let underscoreEmSymbol = Symbol(character: "_")
+    private static let asteriskEmSymbol = Symbol(character: "*")
     private static let underscoreStrongSymbol = Symbol(rawValue: "__")
     private static let asteriskStrongSymbol = Symbol(rawValue: "**")
     private static let tildeStrikethroughSymbol = Symbol(rawValue: "~~")
     private static let equalUnderlineSymbol = Symbol(rawValue: "==")
+    private static let linkTextOpeningSymbol = Symbol(character: "[")
+    private static let linkTextClosingSymbol = Symbol(character: "]")
+    private static let linkURLOpeningSymbol = Symbol(character: "(")
+    private static let linkURLClosingSymbol = Symbol(character: ")")
     
     // MARK: - Static functions
     
-    /// Parses specified string and returns a tuple containing string stripped of tag characters and an array of Markdown elements.
+    /// Parses specified string and returns a tuple containing string stripped of symbols and an array of Markdown elements.
     ///
     /// - Parameter string: String to be parsed.
     /// - Returns: Tuple containing string stripped of tag characters and an array of Markdown elements.
     /// - Throws: Parser error.
     static func parse(_ string: String) throws -> (strippedString: String, elements: [MarkdownElement]) {
         guard
-            let underscoreEmSymbol = underscoreEmSymbol,
-            let asteriskEmSymbol = asteriskEmSymbol,
             let underscoreStrongSymbol = underscoreStrongSymbol,
             let asteriskStrongSymbol = asteriskStrongSymbol,
             let tildeStrikethroughSymbol = tildeStrikethroughSymbol,
             let equalUnderlineSymbol = equalUnderlineSymbol else {
                 return (string, [])
         }
-        
-        func transformToMarkdownElement(_ element: Element) throws -> MarkdownElement {
-            switch element.symbol {
-            case underscoreEmSymbol, asteriskEmSymbol:
-                return .em(element.range)
-            case underscoreStrongSymbol, asteriskStrongSymbol:
-                return .strong(element.range)
-            case tildeStrikethroughSymbol:
-                return .strikethrough(element.range)
-            case equalUnderlineSymbol:
-                return .underline(element.range)
+
+        let underscoreEmRule = Rule(symbol: underscoreEmSymbol)
+        let asteriskEmRule = Rule(symbol: asteriskEmSymbol)
+        let underscoreStrongRule = Rule(symbol: underscoreStrongSymbol)
+        let asteriskStrongRule = Rule(symbol: asteriskStrongSymbol)
+        let tildeStrikethroughRule = Rule(symbol: tildeStrikethroughSymbol)
+        let equalUnderlineRule = Rule(symbol: equalUnderlineSymbol)
+        let linkTextRule = Rule(openingSymbol: linkTextOpeningSymbol, closingSymbol: linkTextClosingSymbol)
+        let linkURLRule = Rule(openingSymbol: linkURLOpeningSymbol, closingSymbol: linkURLClosingSymbol)
+
+        let tokens = try TokenParser.parse(string,
```

Comment:
It might be good to consolidate all the rule declaration into one array in case of manual mistake